pkg/controller/pv_monitor_controller.go

```diff
@@ -0,0 +1,248 @@
+	pvcInformer coreinformers.PersistentVolumeClaimInformer, podInformer coreinformers.PodInformer, nodeInformer coreinformers.NodeInformer) *PVMonitorController {
+
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: client.CoreV1().Events(v1.NamespaceAll)})
+	var eventRecorder record.EventRecorder
+	eventRecorder = broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: fmt.Sprintf("csi-pv-monitor-controller-%s", monitorName)})
+
+	ctrl := &PVMonitorController{
+		csiConn:            conn,
+		eventRecorder:      eventRecorder,
+		supportListVolumes: supportListVolumes,
+		enableNodeWatcher:  enableNodeWatcher,
+		client:             client,
+		monitorName:        monitorName,
+		monitorInterval:    monitorInterval,
+		pvQueue:            workqueue.NewNamed("csi-monitor-pv-queue"),
+		//pvQueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "csi-monitor-pv-queue"),
+
+		pvcToPodsCache: util.NewPVCToPodsCache(),
+		pvEnqueued:     make(map[string]bool),
+	}
+
+	// PV informer
+	pvInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: ctrl.pvAdded,
+		// we do not care about PV changes, so do not need UpdateFunc here.
+		// deleted PVs will not be readded to the queue, so do not need DeleteFunc here
+	})
+	ctrl.pvLister = pvInformer.Lister()
+	ctrl.pvListerSynced = pvInformer.Informer().HasSynced
+
+	// PVC informer
+	ctrl.pvcLister = pvcInformer.Lister()
+	ctrl.pvListerSynced = pvcInformer.Informer().HasSynced
+
+	// Pod informer
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: ctrl.podAdded,
+		// UpdateFunc: ctrl.podUpdated,  TODO: do we need this ?
+		DeleteFunc: ctrl.podDeleted,
+	})
+	ctrl.podLister = podInformer.Lister()
+	ctrl.podListerSynced = podInformer.Informer().HasSynced
+
+	if enableNodeWatcher {
+		ctrl.nodeWatcher = NewNodeWatcher(ctrl.monitorName, ctrl.client, ctrl.pvLister, ctrl.pvcLister, nodeInformer, ctrl.eventRecorder, ctrl.pvcToPodsCache, monitorInterval)
+	}
+
+	ctrl.pvChecker = handler.NewPVHealthConditionChecker(monitorName, conn, client, timeout, ctrl.pvcLister, ctrl.pvLister, ctrl.eventRecorder)
+
+	return ctrl
+}
+
+// Run runs the volume health condition checking method
+func (ctrl *PVMonitorController) Run(workers int, stopCh <-chan struct{}) {
+	defer ctrl.pvQueue.ShutDown()
+
+	klog.Infof("Starting CSI External PV Health Monitor Controller")
+	defer klog.Infof("Shutting down CSI External PV Health Monitor Controller")
+
+	if !cache.WaitForCacheSync(stopCh, ctrl.pvcListerSynced, ctrl.pvListerSynced, ctrl.podListerSynced) {
+		klog.Errorf("Cannot sync cache")
+		return
+	}
+
+	if ctrl.enableNodeWatcher {
+		go ctrl.nodeWatcher.Run(stopCh)
+	}
+
+	if ctrl.supportListVolumes {
```

Comment:
Please add comments here to explain that ListVolumes is preferred for performance reasons.